patches/api/0418-Registry-Modification-API.patch

```diff
@@ -0,0 +1,436 @@
++import org.bukkit.Keyed;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++public final class RegistryListener {
++
++    private RegistryListener() {
++    }
++
++    /**
++     * Called before any object is added to the registry (including ones
++     * from any {@link Freeze} listeners).
++     *
++     * @param <B> builder type
++     */
++    @ApiStatus.OverrideOnly
++    @ApiStatus.Experimental
++    @FunctionalInterface
++    public interface Modification<T extends Keyed, B extends RegistryBuilder<T>> {
++
++        void beforeRegister(@NotNull ModifyRegistryEvent<T, B> event);
++    }
++
++    /**
++     * Called after all other objects have been added to a registry (from
++     * built-in sources or datapacks depending on the registry type).
++     *
++     * @param <T> object type
++     * @param <B> builder type
++     */
++    @ApiStatus.OverrideOnly
++    @ApiStatus.Experimental
++    @FunctionalInterface
++    public interface Freeze<T extends Keyed, B extends RegistryBuilder<T>> {
++
++        void beforeFreeze(@NotNull FreezeRegistryEvent<T, B> event);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/registry/RegistryManager.java b/src/main/java/io/papermc/paper/registry/RegistryManager.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..5788dc7abe8bb13f294dd0bf6f64acf666cdd652
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/RegistryManager.java
+@@ -0,0 +1,17 @@
++package io.papermc.paper.registry;
++
++import org.bukkit.Keyed;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Register listeners to points in a registry's lifecycle.
++ */
++@ApiStatus.NonExtendable
++@ApiStatus.Experimental
++public interface RegistryManager {
++
++    <T extends Keyed, B extends RegistryBuilder<T>> void registerModificationListener(@NotNull RegistryKey2<T> registryKey, @NotNull RegistryListener.Modification<T, B> listener);
++
++    <T extends Keyed, B extends RegistryBuilder<T>> void registerFreezeListener(@NotNull RegistryKey2<T> registryKey, @NotNull RegistryListener.Freeze<T, B> listener);
++}
+diff --git a/src/main/java/io/papermc/paper/registry/RegistryView.java b/src/main/java/io/papermc/paper/registry/RegistryView.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..33939811b0773d8aa85db72ab09990058401a6f7
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/RegistryView.java
+@@ -0,0 +1,20 @@
++package io.papermc.paper.registry;
++
++import net.kyori.adventure.key.Key;
++import net.kyori.adventure.key.Keyed;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Contract;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++@ApiStatus.Experimental
++@ApiStatus.NonExtendable
++public interface RegistryView<T extends Keyed> extends Iterable<T> {
```

Comment:
Because we need a different impl for the `get()` methods. The current impl, just delegates to the registry get methods (which will throw because its not frozen yet). I think this is better than creating different `Registry` instances for different points in the startup. That way the only `Registry` instances created are the final "immutable" ones.

Reply:
couldn't we just have the `get` implementation behave differently before freezing?

Reply:
We could, but the static fields in Registry are filled when this happens, I don't want to support people doing `Registry.GAME_EVENT.get()` inside a listener for the game event registry. Plus, any methods we (or upstream) adds to Registry like anything tag related would also exist in the listeners then. Tags are loaded way later.

Reply:
The name is fine, not permitting accessing the registry outside the provided context is a decent enough reason for this